File: ecosystem/naalp-agent-go/mixedmode/endpoint.go
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.

// Package mixedmode implements N-AALP's mixed-mode / gradual-typing HTTP endpoint SDK
// (design.md Sec.13.8: the mixed-mode discrimination design). A mixed-mode endpoint accepts EITHER a strict
// N-AALP CBOR/COSE envelope OR a legacy NPAMP-CC-HTTP JSON body on the same channel,
// tags each request with its origin, and enforces the one non-negotiable rule the design
// settles before any SDK code: the legacy path gains NO authority the strict checks
// (effect-class, audience, signature) would deny -- it is read-only-equivalent unless the
// caller upgrades to a signed strict object.
//
// This package performs NO cryptography, NO CBOR encoding, and NO carriage-body
// re-serialization of its own: strict-path verification is the real Part-1
// impl/go/envelope.Verify, effect authorization is the real Part-1
// impl/go/policy.Grant.AuthorizeObject, and a legacy body is wrapped octet-for-octet by
// the real Part-1 impl/go/carriage.Carry under ClassHTTP/content_type=json (Sec.13.7) --
// never parsed, never canonicalized, never re-serialized (R-14.4 applies to the legacy
// fallback exactly as it applies to every other carried foreign message). This package's
// own code is only the discrimination order (Sec.13.8(a): try strict first, fall back to
// legacy only on parse failure and only when explicitly enabled) and the ONE
// authorization choke point (Sec.13.8(b)) that caps a legacy-origin record at read_only
// regardless of anything the legacy JSON body itself claims -- an attacker-crafted
// {"effect":3,...} inside an unsigned legacy body is never consulted for authorization;
// Authorize below never even inspects Record.Carriage.Foreign.
package mixedmode

import (
	"github.com/bubblefish-tech/naalp_protocol/impl/go/carriage"
	"github.com/bubblefish-tech/naalp_protocol/impl/go/channels"
	"github.com/bubblefish-tech/naalp_protocol/impl/go/cose"
	"github.com/bubblefish-tech/naalp_protocol/impl/go/envelope"
	"github.com/bubblefish-tech/naalp_protocol/impl/go/policy"
)

// ContentTypeJSON is content_type = 0 (design.md Sec.13.7, the "json" row) -- the
// encoding tag a legacy NPAMP-CC-HTTP body carries when wrapped as a carriage payload.
const ContentTypeJSON uint64 = 0

// ProtocolIDHTTP is protocol_id = 0x03 (design.md Sec.13.4, vectors/registry/protocols.csv
// row 4, "HTTP") -- the registered N-AALP protocol id for HTTP-class carriage, used to
// wrap an accepted legacy body when Endpoint.ProtocolID is left zero.
const ProtocolIDHTTP uint64 = 0x03

// Origin tags a discriminated Record's provenance (Requirement 11 AC2: "a unified typed
// interface that tags each object's origin (legacy vs strict)").
type Origin int

const (
	OriginStrict Origin = iota
	OriginLegacy
)

// String names the origin ("strict" / "legacy").
func (o Origin) String() string {
	if o == OriginStrict {
		return "strict"
	}
	return "legacy"
}

// Error is a named, fail-closed mixedmode error (the same stable-Kind-plus-Msg shape
// every sibling ecosystem package's own glue-layer error type uses, e.g. react.Error).
type Error struct {
	Kind string
	Msg  string
}

func (e *Error) Error() string { return e.Kind + ": " + e.Msg }

// Errors this package's own code can produce. A strict-path verification failure
// propagates as ITS OWN error type (typically *cose.Error), never wrapped into this
// type -- only this package's two glue-layer decisions get a named Error here.
var (
	// ErrLegacyDisabled is returned when the strict parse fails and AllowLegacy is
	// false (AC4: strict is the default; a caller MUST opt in before any legacy
	// fallback is even attempted -- there is no implicit fallback).
	ErrLegacyDisabled = &Error{"LegacyDisabled",
		"endpoint is strict-only; AllowLegacy was not explicitly enabled, so a strict-parse " +
			"failure is refused rather than silently retried as legacy"}

	// ErrLegacyEffectNotAuthorized is the D15 refusal: a legacy-origin record requested
	// an effect above read_only. It is returned unconditionally -- independent of any
	// Grant -- because a legacy record carries no signature-derived principal for a
	// Grant to check against in the first place (design.md Sec.13.8(b)).
	ErrLegacyEffectNotAuthorized = &Error{"LegacyEffectNotAuthorized",
		"a legacy-origin record MUST NOT gain effecting authority above read_only; upgrade " +
			"to a signed strict N-AALP object to request a higher effect (design.md Sec.13.8(b), D15)"}
)

// Record is the unified typed interface Requirement 11 AC2 requires: every
// discriminated request, strict or legacy, decodes into this ONE type, tagged by Origin.
type Record struct {
	// Origin is OriginStrict or OriginLegacy.
	Origin Origin
	// Object is the verified strict N-AALP object -- non-nil iff Origin == OriginStrict.
	Object *envelope.Object
	// Carriage is the legacy body, wrapped octet-for-octet under
	// ClassHTTP/content_type=json (Sec.13.8(a)) -- valid iff Origin == OriginLegacy.
	// Carriage.Foreign holds the raw legacy bytes VERBATIM (R-14.4); this package
	// never parses them, and Authorize never reads this field.
	Carriage carriage.CarriageBody
}

// Endpoint is a mixed-mode N-AALP HTTP endpoint (Requirement 11 AC2-4). The zero value
// is strict-only (AllowLegacy defaults false) -- a caller must set AllowLegacy=true
// explicitly before Handle will ever attempt the legacy fallback (AC4: fail-closed
// opt-in, the legacy path is the wider attack surface).
type Endpoint struct {
	// Profile, Verifier, KindValidator, and KnownCext are passed straight through to
	// the real Part-1 envelope.Verify for the strict-path parse; see
	// impl/go/envelope.Verify's own doc for their meaning.
	Profile       int
	Verifier      cose.Verifier
	KindValidator envelope.KindValidator
	KnownCext     map[uint64]bool

	// AllowLegacy must be set true explicitly for Handle to ever attempt the legacy
	// fallback (AC4).
	AllowLegacy bool

	// ProtocolID is the carriage protocol_id used to wrap an accepted legacy body
	// (design.md Sec.13.4). Zero means ProtocolIDHTTP (0x03).
	ProtocolID uint64

	// SelfAuthority is this endpoint's own consuming-authority id, passed to the real
	// Part-1 envelope.CheckAudience (Sec.2.5.3) for every (channel,kind) registered in
	// ConsumeOnceKinds, on BOTH the strict and legacy legs of Authorize -- mirrors
	// naalp_mixed_mode.MixedModeEndpoint.self_authority in the Python sibling SDK.
	SelfAuthority string

	// ConsumeOnceKinds is the set of (channel, kind) pairs this endpoint treats as
	// single-use/audience-bound (Sec.2.5.3), checked in Authorize on BOTH legs (closes
	// the Fable-identified cross-SDK parity finding, design.md Sec.13.8/D15: this Go
	// mixed-mode SDK previously lacked the audience binding its Python mirror already
	// carried). A strict-origin record targeting one of these pairs MUST additionally
	// pass envelope.CheckAudience against its VERIFIED audience field. A legacy-origin
	// record claiming one of these pairs (named explicitly via Authorize's legacyTarget
	// argument, since Handle never parses the untrusted foreign body) MUST ALSO be
	// refused -- an unsigned legacy record can never present a real audience, so it can
	// never satisfy a consume-once binding (mirrors the Python _handle_legacy surrogate).
	ConsumeOnceKinds map[ConsumeOnceKey]bool
}

// ConsumeOnceKey identifies a (channel, kind) pair by its two wire-registered numeric
// ids (design.md Sec.2.5.3) -- the map key type for Endpoint.ConsumeOnceKinds and the
// explicit legacy-leg target argument to Authorize.
type ConsumeOnceKey struct {
	Channel uint64
	Kind    uint64
}

// Handle discriminates rawBody per design.md Sec.13.8(a): it FIRST attempts to parse
// and verify rawBody as a strict N-AALP envelope (the real Part-1 envelope.Verify, so
// the signature, content id, and every structural check are the genuine ones, never a
// mixedmode-local reimplementation). If and only if that parse fails does Handle
// consider the legacy fallback -- and only when e.AllowLegacy is true; otherwise the
// request is refused ErrLegacyDisabled (AC4 -- no implicit fallback). A legacy body is
// wrapped octet-for-octet as a ClassHTTP/content_type=json carriage payload via the
// real Part-1 carriage.Carry and returned tagged OriginLegacy; it is never parsed as
// JSON by this package (this package makes no assumption about the legacy body's
// structure beyond "it is bytes").
func (e *Endpoint) Handle(rawBody []byte) (*Record, error) {
	obj, err := envelope.Verify(e.Profile, e.Verifier, e.KindValidator, e.KnownCext, rawBody)
	if err == nil {
		return &Record{Origin: OriginStrict, Object: obj}, nil
	}
	if !e.AllowLegacy {
		return nil, ErrLegacyDisabled
	}
	protocolID := e.ProtocolID
	if protocolID == 0 {
		protocolID = ProtocolIDHTTP
	}
	body, cerr := carriage.Carry(protocolID, carriage.ClassHTTP, ContentTypeJSON, nil, "", rawBody)
	if cerr != nil {
		return nil, cerr
	}
	return &Record{Origin: OriginLegacy, Carriage: body}, nil
}

// Authorize is the ONE choke point through which every discriminated Record's
// effecting authority is decided (design.md Sec.13.8(b), D15; requirements.md
// Requirement 11 AC1). It implements the CRITICAL AUTHZ rule: a legacy-origin
// record's effect ceiling is UNCONDITIONALLY read_only, decided by this function
// alone and never derived from anything the legacy body itself contains -- Authorize
// never reads r.Carriage.Foreign, so an attacker-crafted legacy JSON claiming a
// higher effect or a forged audience has nothing here to influence. A strict-origin
// record delegates to the real Part-1 policy.Grant.AuthorizeObject against its
// VERIFIED signer (the exact check every other N-AALP effecting object passes
// through, Sec.6.3) -- no code path in this package routes a legacy-origin record
// around this gate or the channels.CheckEffect choke point on the theory that "it
// isn't really an N-AALP object" (the request-smuggling shape design.md explicitly
// forecloses).
//
// legacyTarget is an OPTIONAL variadic argument (at most its first element is used):
// on the legacy leg it names the (channel, kind) the caller believes this legacy
// record is claiming, since Handle never parses the untrusted foreign body itself
// and a Record's legacy leg therefore carries no channel/kind of its own. Omitting
// it (the zero-arg call every pre-existing caller already makes) is equivalent to a
// legacy claim that never targets a consume-once pair -- no new refusal fires, and
// the D15 read_only ceiling below is unchanged. It is ignored on the strict leg,
// where the (channel, kind) is instead read directly off the VERIFIED r.Object.
func (e *Endpoint) Authorize(r *Record, grant policy.Grant, requestedEffect policy.Effect, legacyTarget ...ConsumeOnceKey) error {
	if r.Origin == OriginLegacy {
		if requestedEffect > policy.ReadOnly {
			return ErrLegacyEffectNotAuthorized
		}
		// (3) the Sec.2.5.3 single-use consume binding, mirrored onto the legacy leg
		// (Python's _handle_legacy surrogate, closing the Fable parity finding). A legacy
		// record has no signature-derived audience -- Handle never parses the foreign
		// body -- so a caller naming a consume-once target here gets a surrogate whose
		// Audience field is left at its Go zero value "" and is NEVER populated from the
		// untrusted foreign bytes (there is no self-asserted-audience field to even read).
		// envelope.CheckAudience's own absent-audience-on-consume-once branch then refuses
		// it unconditionally (ErrWrongAudience): a legacy record can never satisfy a
		// consume-once binding, no matter what (channel, kind) is named.
		if len(legacyTarget) > 0 && e.ConsumeOnceKinds[legacyTarget[0]] {
			surrogate := &envelope.Object{Channel: legacyTarget[0].Channel, Kind: legacyTarget[0].Kind}
			if err := envelope.CheckAudience(surrogate, e.SelfAuthority, true); err != nil {
				return err
			}
		}
		return nil
	}
	// (1) C10 effect-vs-kind admission (design.md Sec.3.2): the object's DECLARED effect must
	// match the effect its kind declares -- the SAME channels.CheckEffect the Python mirror's
	// _authorize_strict runs as its first strict step. A strict object whose declared effect is
	// not the one its kind declares is refused here, before the grant is even consulted; this is
	// the choke point the package doc names, now enforced in-package (not merely delegated).
	if err := channels.CheckEffect(r.Object.Channel, r.Object.Kind, r.Object.Effect); err != nil {
		return err
	}
	// (2) the C5 effect-authorization gate against the VERIFIED signer's registered grant.
	if err := grant.AuthorizeObject(policy.SourceSignature, string(r.Object.Signer), uint64(requestedEffect)); err != nil {
		return err
	}
	// (3) the Sec.2.5.3 single-use consume binding, for a (channel, kind) this endpoint
	// has registered as consume-once -- reuses the real Part-1 envelope.CheckAudience
	// against the VERIFIED strict object and this endpoint's own SelfAuthority (never
	// reimplemented here); closes the Fable-identified parity finding against the
	// Python mirror's _authorize_strict step 3.
	if e.ConsumeOnceKinds[ConsumeOnceKey{Channel: r.Object.Channel, Kind: r.Object.Kind}] {
		if err := envelope.CheckAudience(r.Object, e.SelfAuthority, true); err != nil {
			return err
		}
	}
	return nil
}

File: ecosystem/naalp-mixed-mode/naalp_mixed_mode/endpoint.py
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
"""N-AALP mixed-mode HTTP discrimination endpoint (design.md §13.8, R11).

A `MixedModeEndpoint` is ONE HTTP endpoint that accepts EITHER a strict N-AALP CBOR/COSE
envelope OR a legacy `NPAMP-CC-HTTP` JSON body on the SAME channel, per design.md §13.8(a):
the endpoint FIRST attempts to parse+cryptographically-verify the request body as a strict
N-AALP envelope (a COSE_Sign1-wrapped naalp-object, §2/§4). If and only if that attempt fails
does it fall back to interpreting the body as a `ClassHTTP` (class=1) / `content_type=0` (json)
legacy carriage payload (§13.2, §13.7). No new wire field, CDDL production, or `naalp-version`
bump is introduced -- the discrimination is a parsing-order policy at the endpoint, reusing the
already-registered carriage class/content-type tag.

The mixed-mode discrimination rule (design.md §13.8, R11):
Default STRICT. The legacy/untyped path is an explicit FAIL-CLOSED opt-in that gains NO
authority the strict checks (effect-class, audience, signature) would deny -- a legacy request
carrying no signed effect class + audience MUST NOT gain effecting authority. The legacy path
is read-only-equivalent unless upgraded to a signed strict N-AALP envelope (§2) carrying an
explicit effect class (§6.1, §6.3) and, where consume-once semantics apply, an `audience`
(§2.5.3). No code path here routes a legacy-origin record around the effect-authorization gate
(channels.check_effect / policy.Grant.authorize_object / envelope.check_audience) on the theory
that "it isn't really an N-AALP object" -- a permissive second parser that skipped the strict
authorization checks would be a request-smuggling shape, which this module is built to foreclose:

  * The STRICT path calls the real Part-1 chokepoints unchanged: `channels.check_effect`,
    `policy.Grant.authorize_object` (which in turn calls `policy.resolve_auth_principal` --
    only a signature-derived identity is ever an authorization principal, R-6.5), and, for a
    consume-once (channel, kind), `envelope.check_audience`.
  * The LEGACY path is UNSIGNED by construction, so it can never present a signature-derived
    principal. Any legacy claim of an effect above read_only is routed through the SAME
    `policy.resolve_auth_principal` gate with a non-signature source
    (`policy.SOURCE_FOREIGN_HEADER`), which unconditionally refuses (UnauthenticatedPrincipal)
    -- exactly the same refusal a forged-source strict request would get. And any legacy claim
    targeting a (channel, kind) this endpoint has registered as consume-once is routed through
    the SAME `envelope.check_audience` gate with a surrogate whose `audience` is left absent
    (a self-asserted audience string in the untrusted JSON is NEVER copied onto the surrogate,
    because a claim with no signature behind it carries no authority to bind an audience) --
    `check_audience`'s own absent-audience-on-consume-once branch then raises WrongAudience,
    unconditionally, for every legacy record.

The result: a legacy record can only ever succeed as read_only-equivalent, non-consume-once
carriage. Reaching any higher effect requires actually presenting a signed strict N-AALP
envelope, which takes the strict path from the start.
"""
from . import _bootstrap  # noqa: F401  (repo-relative sys.path setup; must run before `naalp` imports)

import json
from dataclasses import dataclass, field
from typing import Callable, Dict, FrozenSet, Optional, Tuple

from naalp import carriage, cbor, channels, envelope, policy

MODE_STRICT = "strict"
MODE_LEGACY = "legacy"

# §13.4: HTTP carriage's registered standards protocol_id (the reference id this SDK defaults
# a legacy NPAMP-CC-HTTP carriage body to when the caller does not name a specific protocol).
DEFAULT_LEGACY_PROTOCOL_ID = 0x03


class MixedModeError(ValueError):
    """A named, fail-closed mixed-mode endpoint error; .kind is the stable error kind."""

    def __init__(self, kind, msg=""):
        super().__init__("%s: %s" % (kind, msg) if msg else kind)
        self.kind = kind


def default_kind_validator(channel_id: int, kind: int) -> bool:
    """The default strict-path kind validator: REAL dispatch against the frozen twenty-channel
    baseline registry (naalp.channels.lookup), never a permissive `lambda: True`. A (channel,
    kind) pair not registered there is rejected (mirrors envelope.verify's own UnknownKind
    check-order slot, §3.2)."""
    try:
        channels.lookup(channel_id, kind)
        return True
    except channels.UnknownKind:
        return False


@dataclass(frozen=True)
class MixedModeResult:
    """The outcome of one `MixedModeEndpoint.handle()` call. `mode` names which path processed
    the request (MODE_STRICT / MODE_LEGACY). `effect` is the resolved, AUTHORIZED effect class
    (never a self-asserted or unchecked value). `object_` carries the verified strict
    `envelope.Object` (MODE_STRICT only); `carriage_body` carries the wrapped legacy carriage
    body (MODE_LEGACY only). `headers` carries any RFC 9745 Deprecation / RFC 8594 Sunset
    headers a configured `migration.MigrationPolicy` wants attached to the response."""

    mode: str
    effect: int
    object_: Optional["envelope.Object"] = None
    carriage_body: Optional["carriage.CarriageBody"] = None
    headers: Dict[str, str] = field(default_factory=dict)


class MixedModeEndpoint:
    """ONE HTTP endpoint discriminating strict N-AALP envelopes from legacy NPAMP-CC-HTTP JSON
    bodies on the same channel (design.md §13.8).

    `profile`/`alg`/`pubkey` are the fixed verification context for the strict path (a
    mixed-mode endpoint here models one signed peer relationship, matching the SDK-facing scope
    of E1.6 -- not a multi-tenant PKI registry). `kind_validator` defaults to
    `default_kind_validator`. `self_authority` is this endpoint's own consuming-authority id,
    passed to `envelope.check_audience` for both paths. `grants` maps a signer's hex-encoded
    public key to the `policy.Grant` authorizing that signer's ceiling effect; a strict request
    from a signer with no registered grant is refused (EffectNotAuthorized) -- no default,
    ambient authority. `consume_once_kinds` is the set of (channel, kind) pairs this endpoint
    treats as single-use/audience-bound (§2.5.3) on EITHER path. `legacy_protocol_id` is the
    carriage `protocol_id` a recognized legacy body is wrapped under (§13.4).
    `migration_policy` is an optional `migration.MigrationPolicy` whose deprecation headers are
    attached to every result, and which -- once tightened to strict -- makes the legacy path
    refuse outright (LegacyRefused). `allow_legacy` (AC4) is the explicit, fail-closed opt-in
    into the legacy fallback at all: it DEFAULTS FALSE (strict-only), mirroring the Go SDK's
    `Endpoint.AllowLegacy` field (ecosystem/naalp-agent-go/mixedmode/endpoint.go) -- a caller
    must set it True before `handle()` will even ATTEMPT the legacy path on a strict-parse
    failure. There is no implicit fallback: a strict-only endpoint that receives a body its
    strict parser rejects is refused `LegacyDisabled`, never silently retried as legacy."""

    def __init__(
        self,
        *,
        profile: int,
        alg: int,
        pubkey: bytes,
        self_authority: str,
        grants: Dict[str, "policy.Grant"],
        kind_validator: Callable[[int, int], bool] = default_kind_validator,
        consume_once_kinds: FrozenSet[Tuple[int, int]] = frozenset(),
        legacy_protocol_id: int = DEFAULT_LEGACY_PROTOCOL_ID,
        migration_policy=None,
        allow_legacy: bool = False,
    ):
        self._profile = profile
        self._alg = alg
        self._pubkey = bytes(pubkey)
        self._kind_validator = kind_validator
        self._self_authority = self_authority
        self._grants = dict(grants)
        self._consume_once_kinds = frozenset(consume_once_kinds)
        self._legacy_protocol_id = legacy_protocol_id
        self._migration = migration_policy
        self._allow_legacy = allow_legacy

    def _headers(self) -> Dict[str, str]:
        if self._migration is None:
            return {}
        return self._migration.deprecation_headers()

    def handle(self, body: bytes, *, correlation: bytes = b"", method: str = "") -> MixedModeResult:
        """Discriminate and process one request body (§13.8(a)). Attempts the strict parse
        first; ANY failure there (structural, ContentIdMismatch, UnknownKind, an unregistered
        alg, a profile-floor downgrade, or a bad signature -- every failure `envelope.verify`
        itself raises is a `ValueError` subclass, exactly the check-order §13.8(a) calls "that
        parse fails") falls back to the legacy path -- but ONLY when `allow_legacy` was
        explicitly set True at construction (AC4). A strict-only endpoint (the default) that
        cannot parse a body as a strict envelope refuses it outright, `MixedModeError
        ("LegacyDisabled")`, rather than implicitly attempting the legacy interpretation: there
        is no second, permissive parser a caller can reach without opting in. A SUCCESSFUL
        strict parse never falls back: from there, authorization failures (EffectNotAuthorized,
        WrongAudience, ...) propagate as hard fail-closed rejections of the whole request (R11
        AC1's anti-bypass rule -- a validly-signed-but-over-privileged object is REFUSED, never
        silently reinterpreted as a permissive legacy record)."""
        try:
            obj = envelope.verify(self._profile, self._alg, self._pubkey, self._kind_validator, body)
        except ValueError:
            if not self._allow_legacy:
                raise MixedModeError(
                    "LegacyDisabled",
                    "endpoint is strict-only; allow_legacy was not explicitly enabled, so a "
                    "strict-parse failure is refused rather than silently retried as legacy",
                )
            return self._handle_legacy(body, correlation=correlation, method=method)
        return self._authorize_strict(obj)

    def _authorize_strict(self, obj: "envelope.Object") -> MixedModeResult:
        # (1) the object's declared effect must match its kind's C10 declaration (§3.2).
        channels.check_effect(obj.channel, obj.kind, obj.effect)
        # (2) the C5 effect-authorization gate: a signature-derived principal, checked against
        # a REGISTERED grant's ceiling (R-6.3/R-6.5) -- never a default/ambient grant.
        principal = bytes(obj.signer).hex()
        grant = self._grants.get(principal)
        if grant is None:
            raise policy.PolicyError(
                "EffectNotAuthorized", "no grant registered for signer %s" % principal
            )
        grant.authorize_object(policy.SOURCE_SIGNATURE, principal, obj.effect)
        # (3) the §2.5.3 single-use consume binding, for a (channel, kind) this endpoint has
        # registered as consume-once.
        if (obj.channel, obj.kind) in self._consume_once_kinds:
            envelope.check_audience(obj, self._self_authority, consume_once=True)
        return MixedModeResult(
            mode=MODE_STRICT, effect=obj.effect, object_=obj, headers=self._headers()
        )

    def _handle_legacy(self, body: bytes, *, correlation: bytes, method: str) -> MixedModeResult:
        if self._migration is not None and self._migration.state == "strict":
            raise MixedModeError(
                "LegacyRefused",
                "the migration policy has tightened to strict-only; legacy carriage is no "
                "longer accepted on this endpoint",
            )
        try:
            payload = json.loads(body.decode("utf-8"))
        except (json.JSONDecodeError, UnicodeDecodeError) as exc:
            raise MixedModeError(
                "LegacyMalformed", "legacy body is not a strict envelope and is not valid UTF-8 JSON (%s)" % exc
            )
        if not isinstance(payload, dict):
            raise MixedModeError(
                "LegacyMalformed", "legacy NPAMP-CC-HTTP body must be a JSON object"
            )

        # The claimed effect/audience/channel/kind below are SELF-ASSERTED, UNSIGNED JSON --
        # never trusted as authorization input on their own. They are read only to decide
        # whether this record needs to be REFUSED for asking for more than read-only-equivalent
        # carriage; the actual refusal is always produced by the SAME chokepoints the strict
        # path uses (policy.resolve_auth_principal / envelope.check_audience), never by ad hoc
        # comparison logic here (D15's anti-bypass rule).
        try:
            requested_effect = policy.normalize_effect(int(payload.get("effect", policy.READ_ONLY)))
            claimed_channel = payload.get("channel")
            claimed_kind = payload.get("kind")
            if claimed_channel is not None:
                claimed_channel = int(claimed_channel)
            if claimed_kind is not None:
                claimed_kind = int(claimed_kind)
        except (TypeError, ValueError) as exc:
            raise MixedModeError(
                "LegacyMalformed", "legacy body's effect/channel/kind fields are not integers (%s)" % exc
            )
        claimed_method = str(payload.get("method", method or ""))

        carriage_body = carriage.carry(
            protocol_id=self._legacy_protocol_id,
            klass=carriage.CLASS_HTTP,
            content_type=0,  # §13.7: content_type 0 = json
            correlation=correlation,
            method=claimed_method,
            foreign=body,
        )

        if requested_effect > policy.READ_ONLY:
            # D15: an unsigned legacy claim gains NO effecting authority. Force the SAME
            # authorization primitive the strict path's Grant.authorize_object calls; a legacy
            # record has no signature-derived principal, so this call unconditionally refuses
            # (UnauthenticatedPrincipal, R-6.5) -- never a second, permissive parser that skips
            # the gate.
            policy.resolve_auth_principal(policy.SOURCE_FOREIGN_HEADER, "")

        if claimed_channel is not None and claimed_kind is not None:
            key = (claimed_channel, claimed_kind)
            if key in self._consume_once_kinds:
                # D15: the same reasoning applies to the audience-binding gate. Build a
                # surrogate object carrying the UNAUTHENTICATED signer (b"") and, critically,
                # leave `audience` at its default "" -- the JSON's self-asserted `audience`
                # field (if any) is NEVER copied here, because an unsigned claim carries no
                # authority to bind one. envelope.check_audience's own absent-audience-on-
                # consume-once branch then refuses unconditionally (WrongAudience).
                surrogate = envelope.Object(
                    kind=key[1], channel=key[0], signer=b"", created=0,
                    effect=requested_effect, body=cbor.M([]),
                )
                envelope.check_audience(surrogate, self._self_authority, consume_once=True)

        return MixedModeResult(
            mode=MODE_LEGACY, effect=policy.READ_ONLY, carriage_body=carriage_body,
            headers=self._headers(),
        )
